### PR TITLE
fix(screenscraper): let the daily-quota breaker recover

### LINKED ISSUE #4123: [Bug] A single ScreenScraper quota error disables ScreenScraper for an entire scan, and manual matching never recovers
**RomM version**
5.1.1-beta.1. The behaviour dates back to 5.0.0-beta.2 (introduced in #3686), so it
is not specific to recent releases. Re-verified on `master` at `d10bae866`
(2026-09-05), this time by running the code rather than reading it: still present,
all four items. Item 3 is the only part addressed since, and only half of it; see
the note there.

**Describe the bug**
The first time ScreenScraper returns a daily-quota error (HTTP 430 or 431), RomM
sets a flag that makes every subsequent ScreenScraper request fail instantly for the
rest of the scan. It never re-checks. One error, even a transient one or one that has
nothing to do with the user's own allowance, costs them ScreenScraper metadata for
their entire library.

In a real report this tripped **1.4 seconds into the scan**, on the very first ROM:

```
[2026-08-05T11:17:28.517Z] DEBUG [screenscraper] API request: URL=.../jeuInfos.php?crc=919f23cb&...
[2026-08-05T11:17:28.709Z] WARNING [screenscraper] ScreenScraper daily scrape quota exhausted; skipping ScreenScraper for the rest of this scan (quotas reset at midnight CET)
```

Every ROM after that was saved without ScreenScraper data.

Crucially, that user's account was **not** out of quota in the way RomM implied.
Their account page showed 10,010 of 50,000 daily requests used. Their password was
wrong, so ScreenScraper charged their account for every request but never granted
them their member allowance, capping them at the far lower unauthenticated limit.
RomM told them their daily quota was exhausted; their own account page said
otherwise; both were describing different limits.

Four related problems:

**1. HTTP 431 should never stop a scan, and currently does.** 430 and 431 are treated
identically, but they mean completely different things. A 431 tells you two things at
once: this ROM did not match, and the account has reached its daily limit for
submitting unknown ROMs to ScreenScraper's moderation queue for possible addition to
the database.

The first half is ordinary. For metadata purposes it is the same outcome as a plain
"not found", and continuing costs nothing extra: without the cap that request would
have returned 404, so the request count is identical either way.

The second half is a real loss, but not a scraping failure. The user's unmatched ROMs
simply stop being proposed to the moderators for the rest of the day. Every ROM that
does match still scrapes perfectly. Treating this as grounds to disable ScreenScraper
for the entire scan is wrong on both counts.

**2. One error is enough, and there is no recovery.** A single 430 trips the breaker
permanently for that scan. No retry, no backoff, no re-check.

**3. The account check can disable the provider before the scan even starts.**
`prime_account_limits()` runs first and goes through the same request path, so a
quota error there disables ScreenScraper for the whole scan. That call is meant to be
diagnostic and should never be able to switch a provider off.

This half is measured, not theoretical. Running a scan per credential permutation
(#4122) showed `ssuserInfos.php` answering 200 and then refusing the identical request
a second later, which is the developer-credential flapping #4124 documents. The scan
that followed produced no ScreenScraper metadata for any ROM, on an account whose
scraping worked perfectly. PR #4130 has since applied this rule to the credentials
breaker it added: `prime_account_limits()` clears `credentials_rejected` explicitly,
so the account check reports without arming that breaker. The quota breaker got no
equivalent. `get_user_info()` still routes through `_request()`, and the `HTTPException`
branch in `prime_account_limits()` never clears `daily_quota_exhausted`, so a 430 or 431
from `ssuserInfos.php` still takes ScreenScraper out before the scan starts. That is
what remains here.

**4. In the web process the flag is never cleared.** The breaker is only reset by
`begin_scan()`, which runs in the background worker that executes scans. Manual ROM
matching runs in the web process, which never calls it. Once manual matching trips
the flag it stays tripped **indefinitely, including past midnight**, until the
container is restarted. `reset_daily_quota()` exists for exactly this but is no
longer called anywhere outside tests.

**The message is also misleading.** Users are told:

> ScreenScraper daily quota exhausted. It resets at midnight CET.

The user above reasonably checked their ScreenScraper account, saw plenty remaining,
and concluded RomM was broken. Several days went into that. The message also collapses
the two unrelated limits into one string, then asserts the reset time as fact even
when RomM cannot tell which limit was hit or whose it was.

**To Reproduce**

A reliable way to trigger a 430 without exhausting your own allowance:

1. Set `SCREENSCRAPER_PASSWORD` to an incorrect value, leaving the username correct.
2. `docker compose up -d`, then start a scan with ScreenScraper enabled.
3. Scraping still works (ScreenScraper authorizes on RomM's developer credentials),
   but you are capped at the unauthenticated allowance rather than your account's,
   so a 430 arrives quickly.
4. Observe that every remaining ROM is saved without ScreenScraper metadata and no
   further requests are attempted.

For the web-process half: trigger a 430 through manual ROM matching, then wait past
midnight CET. Manual matching still reports the quota as exhausted until RomM is
restarted.

**Expected behavior**

- **HTTP 431 never pauses scanning.** The ROM is recorded as unmatched, exactly as
  with a 404, and the scan continues normally. Every ROM that does match still
  scrapes fine.
- **The submission cap is reported once per scan, at informational level.** Something
  like "ScreenScraper's daily limit for submitting unknown ROMs has been reached;
  unmatched ROMs will not be proposed for review for the rest of today." The user is
  losing a contribution, not a feature, so they should know it happened without it
  being framed as an error.
- More than one 430 is required before ScreenScraper is paused.
- `prime_account_limits()` can never trip the breaker. Any breaker: the rule is that
  the account check reports, and only a request the scan actually needs can take the
  provider out. PR #4130 established this for the credentials breaker; the quota
  breaker still needs it.
- The breaker is temporary. After a backoff, RomM re-checks and resumes if there is
  headroom.
- The breaker resets in the web process too, so manual matching recovers on its own.
- The two limits are described distinctly. HTTP 430 means the daily scrape allowance
  is spent. HTTP 431 means the account has submitted its daily maximum of unrecognized
  ROMs for review, which is not an error and does not indicate anything is wrong with
  the user's setup.

**Additional context**

Reproduced by execution on `master` at `d10bae866`, driving the adapter with a mocked
session the way `tests/adapters/services/test_screenscraper.py` does, so no live
account or live scan is involved. One 431, then three ROMs that would have matched:

```
WARNING ScreenScraper daily unrecognized-ROM quota exhausted; skipping ScreenScraper
        for the rest of this scan (quotas reset at midnight CET)
ROM 1 (unmatched): HTTP 429 -- ScreenScraper daily unrecognized-ROM quota exhausted. It resets at midnight CET.
breaker tripped: True
requests actually sent: 1
ROM 2: short-circuited -- ScreenScraper daily quota exhausted. It resets at midnight CET.
ROM 3: short-circuited -- ScreenScraper daily quota exhausted. It resets at midnight CET.
ROM 4: short-circuited -- ScreenScraper daily quota exhausted. It resets at midnight CET.
requests actually sent after 4 ROMs: 1
```

One request leaves the process for four ROMs. Note the wording as well: a 431 tripped
it, and every ROM after is told the daily *quota* is exhausted.

The same run confirms the other two halves. Calling `reset_scan_state()` then
`prime_account_limits()`, which is what `begin_scan()` does, with `ssuserInfos.php`
answering 430 and again with 431, reports `breaker tripped before the first ROM: True`
both times. And after a 430, `get_user_info()` raises from the guard with no request
leaving the process, which is the re-probe wrinkle described below.

Worth knowing for item 4: `_state` is a module global, so the breaker is per process.
Gunicorn runs `WEB_SERVER_CONCURRENCY` web workers and scans run in a separate RQ
worker, so this should present intermittently, with one web worker refusing manual
matches while another answers normally. Under `DEV_MODE` a scan runs in the web process
(`endpoints/sockets/scan.py:1413`) and resets the breaker there, which is probably why
it does not show up in a dev setup.

Re-checking is free. `ssuserInfos.php` does not consume any quota, verified by
measurement: 150 `jeuInfos.php` calls plus 2 `ssuserInfos.php` calls moved the
account's request counter by exactly 150. A half-open breaker that re-probes after a
backoff costs nothing.

One wrinkle for whoever implements it: the short-circuit sits at the top of `_request()`
(`adapters/services/screenscraper.py:650`), so once the breaker is tripped it refuses
`ssuserInfos.php` too. A half-open re-probe has to bypass that guard rather than simply
calling `get_user_info()`.

Keeping a breaker for 430 is still worthwhile. Once the scrape allowance really is
exhausted, thousands of doomed requests are slow and add avoidable load to
ScreenScraper. The goal is to make it recoverable, not to remove it.

Places that currently assume 431 is fatal and will need updating together. Line
numbers are against `master` at `d10bae866`. They have now shifted twice (PR #4130,
then `24bc4c221`), so treat them as a pointer to the symbol rather than an address:

- `adapters/services/screenscraper.py:317`, which calls `_trip_daily_quota()` on 431,
  identically to the 430 case at `:311`.
- The comment at `adapters/services/screenscraper.py:196-200`, which describes 430 and
  431 as a single "daily request quota".
- `_is_daily_quota_error()` at `handler/metadata/ss_handler.py:244`, which classifies
  any non-rate-limit 429 as a daily quota error and so mislabels 431 to the user.
- `_warn_on_low_quota()` at `adapters/services/screenscraper.py:436`, which raises a
  `log.warning` as the KO allowance runs low. Approaching a submission budget does not
  degrade anything and should not warn.
- `tests/adapters/services/test_screenscraper.py:581`
  (`test_reset_daily_quota_clears_breaker`), which drives the breaker with a 431 and
  asserts it trips, and `:516`
  (`test_request_unrecognized_rom_quota_exhausted_raises_429`), which pins the wording
  that collapses the two limits into one.

Since the last edit, `24bc4c221` added `is_breaker_tripped()` and uses it at
`handler/scan_handler.py:1157` to stop a tripped breaker from clearing the `ss_id` it
never got to check. That works around a consequence rather than the breaker itself, so
everything above still stands, but a fix here should keep that call site correct: it
asks whether the breaker is open, and a half-open one that has since recovered is a
different answer.

Related: #4073 covers manual matching failing even when ScreenScraper is filtered out
of the selected sources, which is the same breaker seen from the UI side.

Minor, same file: `_ScanState.reset()` iterates fields and assigns `f.default`. That
works today because every field has a simple default, but adding a field with
`default_factory` would silently assign the `MISSING` sentinel instead of a real value.

---

*Edit: noted that item 3 is now measured rather than inferred, and that the #4122 fix applies the "the account check never disables the provider" rule to the credentials breaker it adds. The quota breaker is what remains here.*

*Edit 2026-09-04: re-verified against `master` at `2c0fb0879`. Everything here still reproduces. Refreshed the file and line references, which PR #4130 renumbered; recorded that the credentials half of item 3 landed with that PR while the quota half did not; and noted that the tripped breaker also refuses the free `ssuserInfos.php` re-probe.*

*Edit 2026-09-05: re-verified against `master` at `d10bae866`, this time by running the adapter rather than reading it; the transcript is under Additional context. All four items still hold and no fix is pending. Refreshed the line references again (`24bc4c221` shifted this file), and recorded that the same commit now works around a consequence of the breaker in `scan_handler`. Analysis was AI-assisted; the reproduction is mocked at the adapter, not a live scan.*

### PR DESCRIPTION
**Pull request by Spinnich**
**Description**
<sup>Explain the changes or enhancements you are proposing with this pull request.</sup>

Fixes #4123.

The first time ScreenScraper answered HTTP 430 or 431, RomM set a process-wide flag
and every later ScreenScraper request failed instantly without leaving the process.
No threshold, no backoff, no re-check. Only `begin_scan()` cleared it, and that runs
in the RQ scan worker, so a breaker tripped by a manual match in a gunicorn worker
stayed tripped until the container restarted.

Four things change:

**431 stops meaning anything but "not found".** 430 means the daily scrape allowance
is spent; 431 means this ROM did not match *and* the account has proposed its daily
maximum of unknown ROMs to ScreenScraper's moderation queue. Scraping still works
perfectly for every ROM that matches. 431 now returns the empty response a scan
carries on through, and the lost contribution is reported once per scan at info
level. The submission allowance is an order of magnitude smaller than the scrape one
(10,000 vs 100,000 on a tested account), so 431 is the status a normal library hits
first, and it was killing ScreenScraper for entire scans.

**430 arms the breaker only after repeated refusals**, since ScreenScraper answers it
for reasons that do not always survive a retry. Concurrent refusals count once: with
a thread allowance above one, N requests are in flight when the wall is hit and all N
come back 430, which would make any threshold above one meaningless. Each attempt
carries the generation it was sent under, and a refusal only counts if that
generation is still current.

**The breaker is half-open.** Once armed, it re-probes `ssuserInfos.php` (which costs
no quota) every 60 seconds and resumes when the account reports headroom. This is
what lets the web process recover on its own, which is the headline fix: it never
starts a scan, so nothing else was ever going to clear the flag.

**The pre-scan account check can no longer arm anything.** `prime_account_limits()`
now clears the quota breaker in its `except HTTPException` branch, mirroring what
#4130 did for credentials one branch up. The rule is that the account check reports,
and only a request a scan actually needs may take the provider out.

Alongside those, the two cleanups the issue asked for: the low-quota advisory no
longer warns about the submission allowance (running out of it costs a contribution,
not any metadata), and `_ScanState.reset()` honours `default_factory` instead of
writing the `dataclasses._MISSING_TYPE` sentinel over it.

Also addresses #4073. `get_matched_roms_by_name()` was the only ScreenScraper lookup
with no exhausted-provider guard, and `endpoints/search.py` gathers providers without
`return_exceptions`, so a raised ScreenScraper error took the whole manual-match
dialog down with it. It now swallows an exhausted provider the way its three sibling
call sites do. Please check #4073's own repro before closing it.

Related: #4358 (reporting the ROMs a tripped breaker skipped) is deliberately out of
scope here and must not be closed by this PR. It gets cheaper on top of this commit,
since `ss_breaker_tripped` is already the per-ROM signal it needs.

**What to look at hardest**

The recovery path is where the risk is. `_recheck_daily_quota()` deliberately calls
`_attempt_request` rather than `_request`, so it bypasses the guard it is checking;
that bypass is the part worth a careful read.

Second, the `scan_handler.py` change. A recoverable breaker is no longer monotonic
within a scan, which is a property commit 24bc4c221's post-gather `is_breaker_tripped()`
read relied on. With `SCAN_WORKERS > 1`, ROM B's request can now clear the breaker
while ROM A is still gathering its other providers; ROM A then reads `False`, keeps
ScreenScraper in `attempted_sources`, and a complete rescan wipes an `ss_id`
ScreenScraper was never actually asked about. The flag is now sampled inside
`fetch_ss_rom` at entry and in a `finally`, and OR-ed into the existing read. It only
ever discards more from `attempted_sources`, never less, so it cannot wipe an id it
should not. This window is opened by this change, which is why it is in the same
commit.

Third, the concurrency argument in `_recheck_daily_quota()`: the next deadline is
claimed before the first `await`, so read-then-write is atomic on the single-threaded
loop. Same reasoning `RateLimiter.acquire` already documents. If that argument is
wrong, concurrent callers all probe at once.

**Testing**

`TestSubmissionLimit` (4 tests) and `TestDailyQuotaBreaker` (12) in
`tests/adapters/services/test_screenscraper.py`, plus a parametrized priming test and
two manual-match guard tests in `test_ss_handler.py`. Four existing tests that pinned
the broken behaviour were rewritten. Every edge case has a test: 431-then-match, 431
repeated (one line), mixed 430/431, 430 on the retry leg, simultaneous 430s, a success
clearing the count, 430/431 during priming, recovery with headroom, staying armed
without it, probe timeout, probe connection error, both breakers armed, and
`reset_scan_state()` clearing the new state. The tests were written first and run
failing.

The concurrency test was checked for vacuity by deleting the generation check and
confirming `test_simultaneous_refusals_are_one_wall` fails.

346 passed across the ScreenScraper and scan suites; `trunk fmt && trunk check` clean.
No migration, no schema change, no new endpoint, no new i18n keys.

Driven in a browser at both v2 themes with mouse and keyboard, against a live
ScreenScraper account, using a temporary fault-injection harness in the running dev
backend (never committed; reverted before the final `trunk check` and the e2e run):

- Manual match with the breaker armed: HTTP 200, 147 matches rendered, 0 from
  ScreenScraper. With only the new guard reverted, the same request returns HTTP 429
  three times in a row and the whole multi-provider dialog dies, which is #4073.
- The breaker recovering unassisted in the web process, from the backend log:
  `12:18:59 refused 2 requests ... pausing ScreenScraper and re-checking the account
  every 60 seconds` then `12:20:43 the daily scrape quota is available again,
  resuming`. The probe hit the real `ssuserInfos.php`. Observed three times.
- The 431 notice in the Logs view: once, at INFO, with six subsequent 431s producing
  no repeat and no pausing line, and all three searches returning 200.

`npm run test:e2e`: 19 passed, 1 failed. The failure
(`game-media-files.spec.ts` read-only Manual empty state) was verified pre-existing by
running that spec alone on unmodified `upstream/master` @ d10bae866, where it fails
identically. No new e2e spec was added: the behaviour is unreachable from the UI
without fault injection, so a committed spec could only assert the happy path the
existing suite already covers.

**AI assistance**

AI assistance (Claude) was used extensively for this contribution: investigation and
reproduction, code generation for the implementation and its tests, the browser
verification, and drafting this PR description. Every change was reviewed and the
verification above was run by me.

**Checklist**
<sup>Please check all that apply.</sup>

- [x] I've tested the changes locally
- [x] I've updated relevant comments
- [x] I've assigned reviewers for this PR
- [x] I've added unit tests that cover the changes

### CODE REVIEW

File: backend/adapters/services/screenscraper.py [hunk 1]
```diff
@@ -639,25 +714,74 @@ async def _attempt_request(self, url: str, request_timeout: int) -> dict:
...
+            log.debug("ScreenScraper: could not re-check the daily quota (%s)", exc)
+            return False
+
+        # Only a reading this probe brought back is evidence. A body with no
+        # ssuser block leaves the module's limits at their pre-wall value, which
+        # still shows the headroom the account had before it ran out.
+        limits = _state.account_limits
+        if limits is limits_before or (
+            limits is not None and limits.remaining_requests == 0
+        ):
```
[the remark this answers is not in the document]
Comment (Spinnich) [reply backend/adapters/services/screenscraper.py]: Valid, fixed in 63d58ab65.

Confirmed before changing anything: a test driving the probe with `{"response":{"ssuser":{}}}` fails on 0d7ce16d1, and the log says why -- "the daily scrape quota is available again, resuming" off a probe that reported no counters at all. An ssuser block with nothing parseable in it replaces the limits, so the identity check reads it as fresh evidence, and `remaining_requests` is None rather than 0, so the value check passes it too.

Resuming now needs a count that is present and above zero. I kept the identity check rather than replacing it: without it, a body with no ssuser block at all leaves the pre-wall limits in place, and those still show headroom, which is what `test_a_re_check_without_a_reading_stays_armed` pins.

I also considered treating an unreadable reading as permission to resume, on the grounds that an account which never reports counters would otherwise stay armed. Rejected: the probe re-runs every 60s indefinitely, so staying armed is never permanent, while resuming wrongly costs real requests and a false recovery line every minute -- and a wrong resume needs only one flaky response, where a wrong stay-armed needs two independent ones.

Covered by `TestDailyQuotaBreaker::test_a_re_check_with_no_usable_counters_stays_armed`.

This reply was drafted with AI assistance (Claude), as was the fix; both were reviewed by me, and the failing-then-passing test was run locally.

File: backend/handler/scan_handler.py [hunk 2]
```diff
@@ -996,6 +1004,8 @@ async def fetch_ss_rom(playmatch_rom: PlaymatchRomMatch) -> SSRom:
                 note_rate_limited_rom(rom_attrs["fs_name"])
                 attempted_sources.discard(MetadataSource.SS)
                 return SSRom(ss_id=None)
+            finally:
+                ss_breaker_tripped = ss_breaker_tripped or is_breaker_tripped()
```
[the remark this answers is not in the document]
Comment (Spinnich) [reply backend/handler/scan_handler.py:1008]: The mechanism is real, but it is not introduced by this PR and the fix you describe belongs elsewhere.

`git show HEAD~1:backend/handler/scan_handler.py` has `if is_breaker_tripped():` at :1157, from 24bc4c221. That read is process-global too, so on unmodified master ROM B tripping the breaker during ROM A's gather already made A read True and discard ScreenScraper from `attempted_sources`. What this PR adds is two more samples OR-ed into that same read, to close a window a *recoverable* breaker opens (B clearing the breaker mid-gather, which upstream's monotonic breaker could not do). The global-state inference predates it.

The direction of error is deliberate and is called out in the PR body: the samples only ever discard more from `attempted_sources`, never less. The false positive you describe costs a stale ss_id surviving one complete rescan; the false negative it prevents wipes an id ScreenScraper was never asked about. Losing correct metadata is the worse of the two.

Tracking each ROM's own short-circuit is the right long-term answer, and I want it -- it means signalling the short-circuit back out of `ss_handler`'s `_is_provider_exhausted` swallow, which four call sites share (ss_handler.py:878, 1001, 1019, 1051). That is a wider change than this fix, and it is exactly the per-ROM signal #4358 needs (reporting the ROMs a tripped breaker skipped), which already depends on the `ss_breaker_tripped` sample added here. I would rather build it once, there, than approximate it in this PR. I have added this to #4358's scope.

This reply was drafted with AI assistance (Claude); the history claims above were verified by me against the tree.

**Review by gantoine — APPROVED** [review]
